src/Core/Geometry/IndexedGeometry.hpp

```diff
@@ -0,0 +1,491 @@
+    /// \throws std::out_of_range
+    void unlockFirstLayerOccurrence( const LayerSemantic& semanticName );
+
+    //////////////////////////////////////////////////////////////////////
+    //////////////////////////////////////////////////////////////////////
+
+    /// \brief Add layer
+    ///
+    /// Notify observers of the update.
+    /// \return false if a layer with same semantics and name already exists.
+    ///
+    /// \warning Takes the ownership of the layer
+    ///
+    bool addLayer( std::unique_ptr<GeometryIndexLayerBase>&& layer,
+                   const std::string& layerName = "" );
+
+  private:
+    /// Note: we cannot store unique_ptr here has unordered_map needs its
+    /// elements to be copy-constructible
+    using EntryType = std::pair<bool, GeometryIndexLayerBase*>;
+    struct RA_CORE_API KeyHash {
+        std::size_t operator()( const LayerKeyType& k ) const;
+    };
+
+    /// Collection of pairs <lockStatus, Indices>
+    /// \note There is no natural ordering for these elements, thus
+    /// we need an unordered_map. In contrast to map, transparent hashing
+    /// require c++20, so we need to implement them explicitely here
+    /// https://en.cppreference.com/w/cpp/container/unordered_map/find
+    std::unordered_map<LayerKeyType, EntryType, KeyHash> m_indices;
+};
+
+struct RA_CORE_API PointCloudIndexLayer : public GeometryIndexLayer<Vector1ui> {
+    inline PointCloudIndexLayer() : GeometryIndexLayer( staticSemanticName ) {}
+    /// \brief Generate linearly spaced indices with same size as \p attr vertex buffer
+    void generateIndicesFromAttributes( const AttribArrayGeometry& attr );
+
+    static constexpr const char* staticSemanticName = "PointCloud";
+
+  protected:
+    template <class... SemanticNames>
+    inline PointCloudIndexLayer( SemanticNames... names ) :
+        GeometryIndexLayer( staticSemanticName, names... ) {}
+};
+
+struct RA_CORE_API TriangleIndexLayer : public GeometryIndexLayer<Vector3ui> {
+    inline TriangleIndexLayer() : GeometryIndexLayer( staticSemanticName ) {}
+
+    static constexpr const char* staticSemanticName = "TriangleMesh";
+
+  protected:
+    template <class... SemanticNames>
+    inline TriangleIndexLayer( SemanticNames... names ) :
+        GeometryIndexLayer( staticSemanticName, names... ) {}
+};
+
+class RA_CORE_API PolyIndexLayer : public GeometryIndexLayer<VectorNui>
+{
+    inline PolyIndexLayer() : GeometryIndexLayer( "PolyMesh" ) {}
+
+  protected:
+    template <class... SemanticNames>
+    inline PolyIndexLayer( SemanticNames... names ) : GeometryIndexLayer( "PolyMesh", names... ) {}
+};
+
+class RA_CORE_API LineIndexLayer : public GeometryIndexLayer<Vector2ui>
+{
+    inline LineIndexLayer() : GeometryIndexLayer( "LineMesh" ) {}
+
+  protected:
+    template <class... SemanticNames>
+    inline LineIndexLayer( SemanticNames... names ) : GeometryIndexLayer( "LineMesh", names... ) {}
```

Comment:
If I want to generate a LineIndexLayer from another index layer (triangle or poly) the converter should be there ? or an external tranformator ?

Reply:
I prefer external transformator between layers. If a converter between xxxLayer and LineLayer is defined here, all "nice to have" converters should be implemented in all these classes. So, perhaps limiting to "must have" functionality as internal ones is better and user might easily define their own transformator/layer builder. If such transformators are really usefull perhaps we can provide them in a identified namespace ?

Reply:
I also prefer to have an external transformer, which carries the logic of the transformation (e.g. wireframeIndexGenerator ?).